### PR TITLE
Add cookbook for using chained calls for o1 structured outputs

### PR DESCRIPTION
**Pull request by ericning-o**
## Summary

This cookbook describes a few methods for getting structured outputs for o1. 

## Motivation

Currently, o1-preview and o1-mini do not support structured outputs. This may be coming in the future, but developers that use the APIs now currently don't have a defined way to use their existing models that they have for their current infrastructure. 

---

## For new content

When contributing new content, read through our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md), and mark the following action items as completed:

- [x] I have added a new entry in [registry.yaml](https://github.com/openai/openai-cookbook/blob/main/registry.yaml) (and, optionally, in [authors.yaml](https://github.com/openai/openai-cookbook/blob/main/authors.yaml)) so that my content renders on the cookbook website.
- [x] I have conducted a self-review of my content based on the [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md#rubric):
  - [x] Relevance: This content is related to building with OpenAI technologies and is useful to others.
  - [x] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [x] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [x] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [x] Correctness: The information I include is correct and all of my code executes successfully.
  - [x] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{\n",
+      "    \"companies\": [\n",
+      "        {\n",
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:46]: Double check the outputs of this cell. I can see in the markdown version that it outputs the json twice

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      "companies=[CompanyData(company_name='Walmart', page_link='https://en.wikipedia.org/wiki/Walmart', reason='As the largest retailer, Walmart can leverage AI for supply chain optimization, inventory management, personalized customer experiences, and operational efficiency.'), CompanyData(company_name='ExxonMobil', page_link='https://en.wikipedia.org/wiki/ExxonMobil', reason='ExxonMobil can benefit from AI in predictive maintenance, optimizing exploration and drilling, analyzing geological data, and enhancing safety measures.'), CompanyData(company_name='UnitedHealth Group', page_link='https://en.wikipedia.org/wiki/UnitedHealth_Group', reason='UnitedHealth Group can utilize AI to improve patient data analysis, diagnostics, personalize treatment plans, detect fraud, and increase operational efficiency.')]\n"
+     ]
+    }
+   ],
+   "source": [
+    "import requests\n",
+    "from pydantic import BaseModel\n",
+    "from openai import OpenAI\n",
+    "\n",
+    "client = OpenAI()\n",
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:174]: Do we need all of the same code that we had in the previous code cell? Can we reuse that and just apply the 4o chaining on top of it? Might make it a bit cleaner

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    "Read the following html and return which companies would benefit from using AI technology: {html_content}. Rank these propects by opportunity by comparing them and show me the top 3. Return each company with the following information: {CompanyData.model_fields.keys()}\"\n",
+    "\"\"\"\n",
+    "        }\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "o1_response_content = o1_response.choices[0].message.content\n",
+    "\n",
+    "response = client.beta.chat.completions.parse(\n",
+    "    model=\"gpt-4o-2024-08-06\",\n",
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:218]: Would also recommend using 4o-mini for this example. Since this is a relatively simple task, it would be good to showcase the smallest/cheapest model that can do the job, especially if they are pairing this with an expensive model like o1-preview

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    "            \"role\": \"user\", \n",
+    "            \"content\": f\"\"\"\n",
+    "Given the following data, format it with the given response format: {o1_response_content}\n",
+    "\"\"\"\n",
+    "        }\n",
+    "    ],\n",
+    "    response_format=CompaniesData,\n",
+    ")\n",
+    "\n",
+    "print(response.choices[0].message.parsed)"
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:230]: Can you also clean up the json output here and show the schema? Might make it easier to read?

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Conclusion\n",
+    "\n",
+    "The o1 models don't currently have structured outputs support, but we can use the existing structured outputs functionality from `gpt-4o`/`gpt-4o-mini` by chaining two requests together. \n",
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:239]: I would add either here or in the intro a blurb on what the importance of schema adherence is. Why do users want/need it, when it's beneficial, etc. This can help frame the problem for people reading.

Comment (ericning-o) [reply examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:239]: Sounds good!

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb (hunk 1 above)
Comment (ericning-o) [reply examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:46]: Oh I think I just put it in my text manually, I'll just remove it

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb (hunk 3 above)
Comment (ericning-o) [reply examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:218]: Ah yeah, sounds great!

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb (hunk 2 above)
Comment (ericning-o) [reply examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:174]: 🙏

Comment (ericning-o) [reply examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb]: Oh wait, I did this because I don't care about JSON schema adherence in the initial o1 call, so it can simplify the initial prompting.

File: examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+   "metadata": {},
+   "source": [
+    "# Structured Outputs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's now do this with [structured outputs.](https://platform.openai.com/docs/guides/structured-outputs). To enable this functionality, we'll chain our `o1-preview` response, with a `gpt-4o-mini` response that can effectively handle the data returned from the original `o1-preview` response. "
```
Comment (royziv11) [inline examples/o1/Using_chained_calls_for_o1_structured_outputs.ipynb:132]: Minor nit, I would reword this sentence a bit. It reads a bit weird to go: "our `o1-preview` response, with a `gpt-4o-mini` response that can effectively handle the data returned from the original `o1-preview` response."

Should be o1-preview response --> gpt-4o-mini request.

**Comment by royziv11** [comment]
Once you make the small change it should be good to go! Feel free to ship after!

**Review by royziv11 — APPROVED** [review]
